### PR TITLE
Add navigator attestation alternative algorithm

### PR DESCRIPTION
**Pull request by RupertBenWiser**
Took a first swing at trying to write a procedural algorithm for the navigator extension alternative.

### CODE REVIEW

File: docs/alternatives/navigator_extension_alternative.bs [hunk 1]
```diff
@@ -58,10 +58,30 @@ Markup Shorthands: markdown yes, css no
...
+              1. [=Reject=] |promise| and abort these steps.
+          1. Let |attesterSessions| be the [=environment=]'s [=attestations sessions map=]
+          1. Let |domain| be the top level [=origin=]
+          1. Set |session| to |attesterSessions|[|domain|]
+          1. If |session| is <code>null</code> or |session| is timed out then:
+              1. Set |session| to new session from |attester|
+              1. If |session| is <code>null</code> then:
+                  1. [=Reject=] |promise| and abort these steps
+              1. Set |attesterSessions|[|domain|] to |session|
+          1. Set |environmentIntegrityToken| to attestation from |attester| with |contentBinding|
```
Comment (RupertBenWiser) [inline docs/alternatives/navigator_extension_alternative.bs:80]: Really wasn't sure how to write this, would love some guidance.

File: docs/alternatives/navigator_extension_alternative.bs [hunk 2]
```diff
@@ -58,10 +58,30 @@ Markup Shorthands: markdown yes, css no
 
 ### {{Navigator/getEnvironmentIntegrity()}} ### {#navigator-getenvironmentintegrity}
 
-<div class="algorithm" data-algorithm="navigator-getenvironmentintegrity-alg">
+<div algorithm="navigator-getenvironmentintegrity-alg">
+    The browser has a global <dfn for="attestations sessions map">attestation sessions map</dfn>,
+    which is a [=map=] that contains each attestation session. An attestation session contains any
```
Comment (rayankans) [inline docs/alternatives/navigator_extension_alternative.bs:63]: You need to define in a Concept/Infrastructure section what an attestation session is, and its lifetime.

You can also define the methods to interact with it for the purpose of the spec.

Here's an examples: https://www.w3.org/TR/push-api/#push-subscription

If you'd rather iterate on this, that's alright for now but please add a TODO

File: docs/alternatives/navigator_extension_alternative.bs [hunk 3]
```diff
@@ -58,10 +58,30 @@ Markup Shorthands: markdown yes, css no
...
 
-    <i>Todo</i>
+    1. Let |promise| be [=a new promise=]
+    1. Run the following steps [=in parallel=]:
+          1. If |attester| is <code>undefined</code> then:
+              1. [=Reject=] |promise| and abort these steps.
+          1. Let |attesterSessions| be the [=environment=]'s [=attestations sessions map=]
+          1. Let |domain| be the top level [=origin=]
+          1. Set |session| to |attesterSessions|[|domain|]
+          1. If |session| is <code>null</code> or |session| is timed out then:
```
Comment (rayankans) [inline docs/alternatives/navigator_extension_alternative.bs:75]: "timing out" needs to be defines in the infra section as well

File: docs/alternatives/navigator_extension_alternative.bs (hunk 1 above)
Comment (rayankans) [reply docs/alternatives/navigator_extension_alternative.bs:80]: it's easier if you define what methods an 'attester' has exposed, with clear inputs/outputs, and you can call that from the algorithm here.

File: docs/alternatives/navigator_extension_alternative.bs [hunk 4]
```diff
@@ -58,10 +58,30 @@ Markup Shorthands: markdown yes, css no
...
+              1. Set |session| to new session from |attester|
+              1. If |session| is <code>null</code> then:
+                  1. [=Reject=] |promise| and abort these steps
+              1. Set |attesterSessions|[|domain|] to |session|
+          1. Set |environmentIntegrityToken| to attestation from |attester| with |contentBinding|
+          1. If |environmentIntegrityToken| is <code>null</code> then:
```
Comment (rayankans) [inline docs/alternatives/navigator_extension_alternative.bs:81]: let's not use null as an error code here, algorithms can "fail" and you can check if the call you made failed

**Review by rayankans — COMMENTED** [review]
Thanks, left a few comments on how to improve the clarity here

File: docs/alternatives/navigator_extension_alternative.bs (hunk 1 above)
Comment (RupertBenWiser) [reply docs/alternatives/navigator_extension_alternative.bs]: Thanks for the link you gave for how push API does this. Made a lot of these comments a lot easier.

File: docs/alternatives/navigator_extension_alternative.bs (hunk 2 above)
Comment (RupertBenWiser) [reply docs/alternatives/navigator_extension_alternative.bs]: Thanks this made a lot of sense.

I added temporary WebIDLs for what we need and will continue to build on it so I added a TODO.

File: docs/alternatives/navigator_extension_alternative.bs [hunk 5]
```diff
@@ -58,10 +58,30 @@ Markup Shorthands: markdown yes, css no
...
+    The browser has a global <dfn for="attestations sessions map">attestation sessions map</dfn>,
+    which is a [=map=] that contains each attestation session. An attestation session contains any
+    associated state with the attester for an attestation response.
+
     The <dfn method for="Navigator"><code>getEnvironmentIntegrity(|contentBinding|)</code></dfn> method, when invoked, runs these steps:
 
-    <i>Todo</i>
+    1. Let |promise| be [=a new promise=]
+    1. Run the following steps [=in parallel=]:
+          1. If |attester| is <code>undefined</code> then:
```
[the remark this answers is not in the document]
Comment (RupertBenWiser) [reply docs/alternatives/navigator_extension_alternative.bs]: Thanks this is starting to click now, hopefully added this correctly.

File: docs/alternatives/navigator_extension_alternative.bs (hunk 3 above)
Comment (RupertBenWiser) [reply docs/alternatives/navigator_extension_alternative.bs]: Yeah now that we have WebIDL definitions for the attester, it was easy to add a temp method for this concept.

File: docs/alternatives/navigator_extension_alternative.bs (hunk 4 above)
Comment (RupertBenWiser) [reply docs/alternatives/navigator_extension_alternative.bs]: Yeah that makes a lot of sense!

File: docs/alternatives/navigator_extension_alternative.bs [hunk 6]
```diff
@@ -45,6 +45,41 @@ Markup Shorthands: markdown yes, css no
 
 </div>
 
+# Key terms # {#key-terms}
+
+The <dfn for="web environment">web environment</dfn> is defined as <i>TODO</i>
+
+# Attesters # {#attester}
+
+The term <dfn for="attester">attester</dfn> refers to a third party capable of returning an Web
```
Comment (rayankans) [inline docs/alternatives/navigator_extension_alternative.bs:54]: nit: a Web

File: docs/alternatives/navigator_extension_alternative.bs [hunk 7]
```diff
@@ -45,6 +45,41 @@ Markup Shorthands: markdown yes, css no
...
+Environment Integrity verdict. A <dfn for="verdict">Web Environment Integrity verdict</dfn> refers
```
Comment (rayankans) [inline docs/alternatives/navigator_extension_alternative.bs:55]: nit: Make `Environment Integrity verdict` linkable

File: docs/alternatives/navigator_extension_alternative.bs [hunk 8]
```diff
@@ -45,6 +45,41 @@ Markup Shorthands: markdown yes, css no
...
+## {{AttesterConnection}} ## {#attester-connection}
+
+<xmp class="idl">
+  interface AttesterConnection {
+    ArrayBuffer getAttestation(DOMString contentBinding);
+  };
+</xmp>
+
+: getAttestation
+:: Returns a COSE signed CBOR object as an ArrayBuffer from the [=attester=] that contains the [=verdict=].
```
Comment (rayankans) [inline docs/alternatives/navigator_extension_alternative.bs:81]: is this verdict the `Environment Integrity verdict`?

Comment (RupertBenWiser) [reply docs/alternatives/navigator_extension_alternative.bs:81]: Yeah, updated to make the reference the exact same field for clarity. Also updated the name to Integrity verdict to make it less of a mouthful.

File: docs/alternatives/navigator_extension_alternative.bs (hunk 7 above)
Comment (RupertBenWiser) [reply docs/alternatives/navigator_extension_alternative.bs:55]: Interesting, I had trouble linking this when the dfn was just `verdict`, but it works fine when it is `Integrity verdict`.

**Review by rayankans — APPROVED** [review]

**Pull request merged**